### ISSUE TITLE
different number of loci when converting to genind object

### ISSUE THREAD

**Issue by mcasso85** [thibautjombart/adegenet#101]
I get this (see example below):
First, a number of loci when read.vcf
And, when converting with loci2genind I get almost the double number of loci.

I've tried with different datasets (vcf from different analyses, and/or with different raw data), and I always get the same. Only in one case, I have 50 loci when read.vcf and I get 50 loci after converting with loci2genind. I think that the problem happens when I have missing data (SNPs with ./. ). Is it possible?

Any tip will be veery appreciated! Thanks!

Maria.

Example:
> data.loci <- read.vcf("mydata.vcf")
Reading 983 / 983 loci.
Done.
> data.genind <- loci2genind(data.loci)
Warning message:
In df2genind(as.matrix(x[, attr(x, "locicol")]), sep = "[/\\|]",  :
  duplicate labels detected for some loci; using generic labels
> data.genind
/// GENIND OBJECT /////////

 // 6 individuals; 1,966 loci; 2,949 alleles; size: 853.7 Kb

 // Basic content
   @tab:  6 x 2949 matrix of allele counts
   @loc.n.all: number of alleles per locus (range: 1-2)
   @loc.fac: locus factor for the 2949 columns of @tab
   @all.names: list of allele names for each locus
   @ploidy: ploidy of each individual  (range: 2-2)
   @type:  codom
   @call: df2genind(X = as.matrix(x[, attr(x, "locicol")]), sep = "[/\\|]", 
    pop = pop)

 // Optional content
   - empty -
>

**Comment by romunov** [comment]
Can you provide the original file, replicate the issue with one of datasets that comes with any CRAN package or if do a `dput` of `data.loci`?

**Comment by mcasso85** [comment]
Hello!
I subsetted my data to 21 loci (rows 10 to 30 in the vcf). To read the vcf file, my code is:
```
> data.loci <- read.vcf("mydata.vcf")
Reading 20 / 20 loci.
Done.
```
This is a new issue! I know there are 21 loci...


Then I convert it to genind object and I get 38 loci:
```
> data <- loci2genind(data.loci)
Warning message:
In df2genind(as.matrix(x[, attr(x, "locicol")]), sep = "[/\\|]",  :
  duplicate labels detected for some loci; using generic labels
> data
/// GENIND OBJECT /////////

 // 6 individuals; 38 loci; 58 alleles; size: 21.2 Kb

 // Basic content
   @tab:  6 x 58 matrix of allele counts
   @loc.n.all: number of alleles per locus (range: 1-2)
   @loc.fac: locus factor for the 58 columns of @tab
   @all.names: list of allele names for each locus
   @ploidy: ploidy of each individual  (range: 2-2)
   @type:  codom
   @call: df2genind(X = as.matrix(x[, attr(x, "locicol")]), sep = "[/\\|]", 
    pop = pop)

 // Optional content
   - empty -
```


Here is the `dput`:
```
> dput(data.loci)
structure(list(`7` = structure(c(1L, 1L, 1L, 1L, 2L, 3L), .Label = c("A/A", 
"A/T", "./."), class = "factor"), `7` = structure(c(1L, 1L, 2L, 
2L, 3L, 2L), .Label = c("A/A", "./.", "A/G"), class = "factor"), 
    `7` = structure(c(1L, 1L, 1L, 1L, 2L, 3L), .Label = c("A/A", 
    "A/G", "./."), class = "factor"), `7` = structure(c(1L, 1L, 
    1L, 2L, 1L, 3L), .Label = c("T/T", "C/T", "./."), class = "factor"), 
    `7` = structure(c(1L, 1L, 1L, 2L, 1L, 3L), .Label = c("C/C", 
    "A/C", "./."), class = "factor"), `11` = structure(c(1L, 
    2L, 1L, 1L, 1L, 1L), .Label = c("C/C", "C/T"), class = "factor"), 
    `11` = structure(c(1L, 2L, 1L, 1L, 1L, 1L), .Label = c("T/T", 
    "A/T"), class = "factor"), `66` = structure(c(1L, 2L, 3L, 
    1L, 4L, 3L), .Label = c("A/G", "./.", "G/G", "A/A"), class = "factor"), 
    `70` = structure(c(1L, 1L, 1L, 2L, 2L, 3L), .Label = c("./.", 
    "A/G", "G/G"), class = "factor"), `70` = structure(c(1L, 
    1L, 1L, 2L, 2L, 3L), .Label = c("./.", "C/G", "G/G"), class = "factor"), 
    `74` = structure(c(1L, 2L, 1L, 2L, 3L, 1L), .Label = c("T/T", 
    "./.", "A/T"), class = "factor"), `74` = structure(c(1L, 
    2L, 3L, 2L, 4L, 3L), .Label = c("C/G", "./.", "C/C", "G/G"
    ), class = "factor"), `74` = structure(c(1L, 2L, 3L, 1L, 
    4L, 3L), .Label = c("A/G", "./.", "G/G", "A/A"), class = "factor"), 
    `78` = structure(c(1L, 1L, 1L, 2L, 1L, 3L), .Label = c("T/T", 
    "C/T", "./."), class = "factor"), `78` = structure(c(1L, 
    1L, 1L, 2L, 1L, 3L), .Label = c("G/G", "A/G", "./."), class = "factor"), 
    `78` = structure(c(1L, 2L, 3L, 1L, 4L, 2L), .Label = c("C/T", 
    "./.", "T/T", "C/C"), class = "factor"), `78` = structure(c(1L, 
    1L, 1L, 2L, 1L, 3L), .Label = c("T/T", "C/T", "./."), class = "factor"), 
    `78` = structure(c(1L, 2L, 2L, 2L, 3L, 4L), .Label = c("C/T", 
    "T/T", "C/C", "./."), class = "factor"), `78` = structure(c(1L, 
    2L, 1L, 3L, 1L, 3L), .Label = c("A/A", "A/G", "./."), class = "factor"), 
    `117` = structure(c(1L, 2L, 1L, 1L, 1L, 3L), .Label = c("G/G", 
    "A/G", "./."), class = "factor"), `117` = structure(c(1L, 
    2L, 1L, 1L, 1L, 3L), .Label = c("T/T", "A/T", "./."), class = "factor")), .Names = c("7", 
"7", "7", "7", "7", "11", "11", "66", "70", "70", "74", "74", 
"74", "78", "78", "78", "78", "78", "78", "117", "117"), class = c("loci", 
"data.frame"), locicol = 1:21, row.names = c("IND01", "IND02", 
"IND03", "IND04", "IND05", "IND06"))
```

**Comment by thibautjombart** [comment]
OK I'm looking into it now.. but for what it's worth, there's already an issue: loci names should be unique, while I get:
```r
> x
        7   7   7   7   7  11  11  66  70  70  74  74  74  78  78  78  78  78
IND01 A/A A/A A/A T/T C/C C/C T/T A/G ./. ./. T/T C/G A/G T/T G/G C/T T/T C/T
IND02 A/A A/A A/A T/T C/C C/T A/T ./. ./. ./. ./. ./. ./. T/T G/G ./. T/T T/T
IND03 A/A ./. A/A T/T C/C C/C T/T G/G ./. ./. T/T C/C G/G T/T G/G T/T T/T T/T
IND04 A/A ./. A/A C/T A/C C/C T/T A/G A/G C/G ./. ./. A/G C/T A/G C/T C/T T/T
IND05 A/T A/G A/G T/T C/C C/C T/T A/A A/G C/G A/T G/G A/A T/T G/G C/C T/T C/C
IND06 ./. ./. ./. ./. ./. C/C T/T G/G G/G G/G T/T C/C G/G ./. ./. ./. ./. ./.
       78 117 117
IND01 A/A G/G T/T
IND02 A/G A/G A/T
IND03 A/A G/G T/T
IND04 ./. G/G T/T
IND05 A/A G/G T/T
IND06 ./. ./. ./.
> 
```

**Comment by thibautjombart** [comment]
OK, so it turns out this is a problem in loci2genind, which should treat "." as a missing data and does not. I have just fixed it in my fork of pegas and made a pull request:
https://github.com/emmanuelparadis/pegas/pull/15
@emmanuelparadis this is a tiny change but should fix conversion issues.

Now, I still need to investigate why alleles called "." causes more loci to appear in the constructor.

**Comment by thibautjombart** [comment]
@mcasso85 waiting for the pull request to be merged, can you can try:
```
library(devtools)
install_github("thibautjombart/pegas")
```
and try again importing your data?

**Comment by thibautjombart** [comment]
Right, I got to the bottom of it. So, what was happening was:
* missing data in loci object were "."
* they were not declared as such when converting data using df2genind; as a result, they were treated as alleles
* the constructor of `genind` objects expect a table with columns named as [locus].[alleles]; so if there is a "." in the alleles or in the loci that messes everything up; I now added checks for this kind of situation, both in `df2genind` and in the constructor

**Comment by thibautjombart** [comment]
Fixed in adegenet as of 2b6cf5dbb43f7c20d9e8c90a75568033d8458c8d; will need to patch for pegas.
